clang/lib/CAS/IncludeTree.cpp

```diff
@@ -222,29 +240,44 @@ IncludeTree::FileList::getFileSize(size_t I) const {
 llvm::Error IncludeTree::FileList::forEachFile(
     llvm::function_ref<llvm::Error(File, FileSizeTy)> Callback) {
   size_t I = 0;
+  size_t FileCount = getNumFilesCurrentList();
   return forEachReference([&](ObjectRef Ref) -> llvm::Error {
-    auto Include = getFile(Ref);
-    if (!Include)
-      return Include.takeError();
-    return Callback(std::move(*Include), getFileSize(I++));
+    if (I < FileCount) {
+      auto Include = getFile(Ref);
+      if (!Include)
+        return Include.takeError();
+      return Callback(std::move(*Include), getFileSize(I++));
+    }
+    // Otherwise, it's a chained FileList.
+    ++I;
+    auto Proxy = getCAS().getProxy(Ref);
+    if (!Proxy)
+      return Proxy.takeError();
+    FileList FL(std::move(*Proxy));
+    return FL.forEachFile(Callback);
```

Comment:
Could you enhance `FileList::forEachFile()` and have it skip a `FileList` that was already visited?

Reply:
SGTM. Also easy to skip visiting individual files that are identical and contained in multiple lists.

Reply:
Fixed in 2066527eca2d